File: contains-duplicate-iii/contains-duplicate-iii.java
class Solution {
    public boolean containsNearbyAlmostDuplicate(int[] nums, int k, int t) {
        if (k < 1 || t < 0) {
            return false;
        }
            
        TreeMap<Long, Integer> tm = new TreeMap<Long, Integer>(); // Counter {key : count}
        
        int size = nums.length;
        for (int i = 0; i < size; ++i) {
            Long n = new Long(nums[i]);
            
            // Remove out of bound key
            if (i - k > 0) {
                Long v = new Long(nums[i-k-1]);
                int c = tm.get(v);
                if (c == 1) {
                    tm.remove(v);
                } else {
                    tm.put(v, c - 1);
                }             
            }
            
            // Compare with closest value in map
            Long lo = tm.floorKey(n);
            if (lo != null && Math.abs(n - lo) <= t) {
                return true;           
            }
            Long hi = tm.ceilingKey(n);
            if (hi != null && Math.abs(n - hi) <= t) {
                return true;
            }
            
            // Put n
            tm.put(n, tm.getOrDefault(n, 0) + 1);
        }
        return false;
    }
}

File: 220-contains-duplicate-iii/contains-duplicate-iii.py
class Solution:
    def containsNearbyAlmostDuplicate(self, nums: List[int], k: int, t: int) -> bool:
        if t < 0:
            return False  # edge case

        buckets = {}
        for i, num in enumerate(nums):
            bucket_key = num // (t + 1)
            if bucket_key in buckets and i - buckets[bucket_key][0] <= k:
                return True
            if bucket_key - 1 in buckets and i - buckets[bucket_key - 1][0] <= k and abs(num - buckets[bucket_key - 1][1]) <= t:
                return True
            if bucket_key + 1 in buckets and i - buckets[bucket_key + 1][0] <= k and abs(num - buckets[bucket_key + 1][1]) <= t:
                return True
            buckets[bucket_key] = (i, num)
        return False